designs/extensible-cli-and-scaffolding-plugins-phase-1.md

```diff
@@ -0,0 +1,206 @@
+# Extensible CLI and Scaffolding Plugins
+
+## Overview
+
+I would like for Kubebuilder to become more extensible, such that it could be imported and used as a library in other projects. Specifically, I'm looking for a way to use Kubebuilder's existing CLI and scaffolding for Go projects, but to also be able to augment the Kubebuilder project structure with other custom project types so that I can support the Kubebuilder workflow with non-Go operators (e.g. operator-sdk's Ansible and Helm-based operators).
+
+The idea is for Kubebuilder to define one or more plugin interfaces that can be used to drive what the `init`, `create api` and `create webhooks` subcommands do and to add a new `cli` package that other projects can use to integrate out-of-tree plugins with the Kubebuilder CLI in their own projects.
+
+## Related issues and PRs
+
+* [#1148](https://github.com/kubernetes-sigs/kubebuilder/pull/1148)
+* [#1171](https://github.com/kubernetes-sigs/kubebuilder/pull/1171)
+* Possibly [#1218](https://github.com/kubernetes-sigs/kubebuilder/issues/1218)
+
+## Prototype implementation
+
+Barebones plugin refactor: https://github.com/joelanford/kubebuilder-exp
+Kubebuilder feature branch: https://github.com/kubernetes-sigs/kubebuilder/tree/feature/plugins-part-2-electric-boogaloo
+
+## Plugin interfaces
+
+### Required
+
+Each plugin would minimally be required to implement the `Plugin` interface.
+
+```go
+type Plugin interface {
+    // Version returns the plugin's semantic version, ex. "v1.2.3".
+    //
+    // Note: this version is different from config version.
+    Version() string
+    // Name returns a DNS1123 label string defining the plugin type.
+    // For example, Kubebuilder's main plugin would return "go".
+    //
+    // Plugin names can be fully-qualified, and non-fully-qualified names are
+    // prepended to ".kubebuilder.io" to prevent conflicts.
+    Name() string
+    // SupportedProjectVersions lists all project configuration versions this
+    // plugin supports, ex. []string{"2", "3"}. The returned slice cannot be empty.
+    SupportedProjectVersions() []string
```

Comment:
After looking over our discussion notes, I realized that `Plugin.Version()` and a `PROJECT` file's `version` field can/should drift since the former describes a plugin's development cycle while the latter describes the overall project's. Furthermore, indexing the CLI's plugins, which are indexed by semantic version, with a single digit version string (`version: "2"`) will not work, especially due to the development cycle disconnect described above.

The `SupportedProjectVersions()` allow a plugin to be added to the CLI's plugins such that they can be indexed by project version from `PROJECT`/`--project-version`:

```go
func WithPlugins(plugins ...plugin.Base) Option {
	return func(c *cli) error {
		for _, p := range plugins {
			for _, ver := range p.SupportedProjectVersions() {
				if _, ok := c.plugins[ver]; !ok {
					c.plugins[ver] = []plugin.Base{}
				}
				c.plugins[ver] = append(c.plugins[ver], p)
			}
		}
		return nil
	}
}
```

@DirectXMan12 @mengqiy @joelanford @Adirio do my concerns hold water/solution make sense?

Reply:
+1 on that.

Reply:
+1 seems fine to me

Reply:
So basically each plugin will have a method that returns which versions it should be added to. That sounds ok to me.